src/Network/Transport/TCP.hs

```diff
@@ -1753,6 +1765,26 @@ withScheduledAction ourEndPoint f =
           (traverse (\(tp, a) -> runScheduledAction (ourEndPoint, tp) a) <=< readIORef)
           (\ref -> f (\rp g -> mask_ $ schedule rp g >>= \x -> writeIORef ref (Just (rp,x)) ))
 
+-- | Run an IO but kill it after an optional timeout in microseconds.
+withTimeout :: Exception e => e -> Int -> IO t -> IO t
+withTimeout e timeout io = mask $ \unmask -> do
+  var <- newEmptyMVar
+  rec {
+      tidTimeout <- forkIO $ do
+        unmask (threadDelay timeout >> killThread tidAction)
+        putMVar var Nothing
+    ; tidAction  <- forkIO $ do
+        -- If there's an exception, it'll be thrown up top when we takeMVar
+        -- and pattern-match on the value within.
+        outcome <- unmask $
+          (io `catch` (throw :: SomeException -> IO t)) <* killThread tidTimeout
```

Comment:
Shouldn't `withTimeout` terminate immediately if `io` throws an exception?

How is this function better than `System.Timeout.timeout`?